# Assignment: Hypothesis Testing with Z-Test and t-Test
### Left-tailed, right-tailed and two-tailed tests on a student dataset

**Total marks:** 100  **Submit:** one `.ipynb` file

**What you will do**

You are given a small dataset of **40 college students**. Answer **6 questions**:

| Question | Test | Tail |
|---|---|---|
| Q1 | Z-test | Left |
| Q2 | Z-test | Right |
| Q3 | Z-test | Two |
| Q4 | t-test | Left |
| Q5 | t-test | Right |
| Q6 | t-test | Two |

Then fill in the summary table in Q7.

**Instructions**
1. Run the two starter cells first. **Do not change them.**
2. Write your code only in the cells marked `YOUR CODE HERE`.
3. Write your answers in the cells marked **Your answer**.
4. Use **α = 0.05** for every question.
5. Allowed libraries: `numpy`, `pandas`, `scipy.stats`, `matplotlib`.
6. Compute every number from the dataset. Do not type in the answers by hand.

## Quick Reference

| | Z-test | t-test |
|---|---|---|
| Use when | Population σ is **known** | Population σ is **unknown** |
| Formula | $z = \dfrac{\bar{x} - \mu_0}{\sigma / \sqrt{n}}$ | $t = \dfrac{\bar{x} - \mu_0}{s / \sqrt{n}}$, $df = n - 1$ |

| Tail | $H_1$ | p-value (Z) | p-value (t) | Critical value (Z, α = 0.05) |
|---|---|---|---|---|
| Left | $\mu < \mu_0$ | `stats.norm.cdf(z)` | `stats.t.cdf(t, dof)` | −1.645 |
| Right | $\mu > \mu_0$ | `1 - stats.norm.cdf(z)` | `1 - stats.t.cdf(t, dof)` | +1.645 |
| Two | $\mu \neq \mu_0$ | `2 * (1 - stats.norm.cdf(abs(z)))` | `2 * (1 - stats.t.cdf(abs(t), dof))` | ±1.960 |

**Decision rule:** if **p-value ≤ α**, reject $H_0$. Otherwise, fail to reject $H_0$.

For the sample standard deviation use `df['column'].std()`. pandas uses `ddof=1` by default, which is what the t-test needs.

⚠️ The dataset is stored in a variable called `df`. Store the degrees of freedom in a variable called **`dof`**, not `df`, or you will overwrite the dataset.

## Setup and Dataset

In [3]:
# ---------------- STARTER CODE – DO NOT MODIFY ----------------
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy import stats

alpha = 0.05
print("Libraries loaded. alpha =", alpha)

Libraries loaded. alpha = 0.05


In [4]:
# ---------------- STARTER CODE – DO NOT MODIFY ----------------
# Dataset: 40 students from three sections of a college course
df = pd.DataFrame({
    'student_id':  [f'S{i:02d}' for i in range(1, 41)],
    'section':     ['A'] * 14 + ['B'] * 13 + ['C'] * 13,
    'sleep_hours': [6.9, 5.5, 7.4, 7.6, 4.5, 5.2, 6.7, 6.3, 6.6, 5.7, 7.6, 7.5, 6.7, 7.8,
                    7.1, 5.7, 7.0, 5.5, 7.6, 6.5, 6.4, 5.9, 7.9, 6.4, 6.1, 6.2, 7.2,
                    7.0, 7.1, 7.1, 9.0, 6.2, 6.0, 5.7, 7.3, 7.8, 6.5, 5.7, 5.7, 7.3],
    'screen_time': [6.5, 6.2, 4.5, 5.8, 5.6, 5.8, 6.7, 5.8, 6.4, 5.5, 5.9, 6.3, 3.4, 5.0,
                    4.8, 4.6, 5.1, 7.5, 4.2, 6.8, 3.1, 5.0, 5.7, 6.3, 6.4, 6.6, 5.0,
                    4.8, 6.7, 5.2, 3.7, 3.9, 4.2, 6.1, 5.6, 6.4, 4.9, 5.7, 6.3, 5.0],
    'exam_score':  [71, 60, 63, 63, 55, 71, 62, 67, 71, 71, 73, 66, 62, 66,
                    50, 52, 53, 57, 70, 57, 63, 79, 63, 74, 57, 64, 57,
                    63, 75, 49, 71, 69, 61, 52, 67, 61, 69, 67, 83, 64],
    'attendance':  [75, 82, 82, 89, 86, 83, 90, 74, 77, 75, 79, 73, 85, 80,
                    72, 75, 83, 86, 93, 98, 83, 75, 68, 83, 76, 79, 77,
                    80, 87, 82, 80, 75, 71, 78, 81, 92, 82, 87, 78, 74],
    'study_hours': [2.2, 2.4, 4.7, 2.3, 3.7, 2.3, 3.7, 3.3, 2.9, 3.0, 2.5, 3.4, 2.6, 2.0,
                    2.0, 3.1, 4.3, 3.1, 2.9, 3.2, 4.0, 3.2, 2.7, 3.9, 3.3, 4.2, 3.1,
                    2.0, 1.9, 4.3, 4.4, 2.9, 2.7, 4.2, 2.1, 2.3, 3.5, 2.7, 3.0, 2.9],
    'commute_min': [34, 42, 33, 37, 18, 32, 26, 23, 26, 30, 38, 23, 32, 29,
                    30, 39, 36, 41, 31, 27, 30, 34, 33, 24, 33, 34, 50,
                    45, 26, 30, 22, 28, 34, 40, 27, 27, 17, 31, 25, 28],
})

print("Shape:", df.shape)
print(df['section'].value_counts().sort_index().to_string())
df.head(10)

Shape: (40, 8)
section
A    14
B    13
C    13


,student_id,section,sleep_hours,screen_time,exam_score,attendance,study_hours,commute_min
0,S01,A,6.9,6.5,71,75,2.2,34
1,S02,A,5.5,6.2,60,82,2.4,42
2,S03,A,7.4,4.5,63,82,4.7,33
3,S04,A,7.6,5.8,63,89,2.3,37
4,S05,A,4.5,5.6,55,86,3.7,18
5,S06,A,5.2,5.8,71,83,2.3,32
6,S07,A,6.7,6.7,62,90,3.7,26
7,S08,A,6.3,5.8,67,74,3.3,23
8,S09,A,6.6,6.4,71,77,2.9,26
9,S10,A,5.7,5.5,71,75,3.0,30


**Data dictionary**

| Column | Meaning |
|---|---|
| `student_id` | Student ID |
| `section` | Class section: A (14 students), B (13), C (13) |
| `sleep_hours` | Average sleep per night (hours) |
| `screen_time` | Average daily screen time (hours) |
| `exam_score` | Final exam score (out of 100) |
| `attendance` | Attendance (%) |
| `study_hours` | Average self-study per day (hours) |
| `commute_min` | One-way commute time (minutes) |

Explore the data before you start (optional):

In [5]:
# ===================== OPTIONAL – EXPLORE THE DATA =====================
# OPTIONAL: explore the dataset, e.g. df.describe() or df.groupby('section').mean(numeric_only=True)
print("Dataset Shape:")
print(df.shape)

print("\nColumn Names:")
print(df.columns.tolist())

print("\nBasic Statistical Summary:")
display(df.describe())

print("\nSection-wise Student Count:")
print(df['section'].value_counts().sort_index())

print("\nSection-wise Mean of Numerical Variables:")
display(df.groupby('section').mean(numeric_only=True))

print("\nMissing Values in Each Column:")
print(df.isnull().sum())

Dataset Shape:
(40, 8)

Column Names:
['student_id', 'section', 'sleep_hours', 'screen_time', 'exam_score', 'attendance', 'study_hours', 'commute_min']

Basic Statistical Summary:


,sleep_hours,screen_time,exam_score,attendance,study_hours,commute_min
count,40.000000,40.00000,40.000000,40.000000,40.000000,40.000000
mean,6.647500,5.47500,64.200000,80.625000,3.072500,31.125000
std,0.900993,1.01571,7.855098,6.463656,0.761236,6.962123
min,4.500000,3.10000,49.000000,68.000000,1.900000,17.000000
25%,5.975000,4.87500,59.250000,75.000000,2.475000,26.750000
50%,6.650000,5.65000,63.500000,80.000000,3.000000,30.500000
75%,7.300000,6.30000,70.250000,83.500000,3.550000,34.000000
max,9.000000,7.50000,83.000000,98.000000,4.700000,50.000000



Section-wise Student Count:
section
A    14
B    13
C    13
Name: count, dtype: int64

Section-wise Mean of Numerical Variables:


,sleep_hours,screen_time,exam_score,attendance,study_hours,commute_min
section,,,,,,
A,6.571429,5.671429,65.785714,80.714286,2.928571,30.214286
B,6.576923,5.469231,61.230769,80.615385,3.307692,34.000000
C,6.800000,5.269231,65.461538,80.538462,2.992308,29.230769



Missing Values in Each Column:
student_id     0
section        0
sleep_hours    0
screen_time    0
exam_score     0
attendance     0
study_hours    0
commute_min    0
dtype: int64


---
# Part A – Z-Tests (σ known, use all 40 students)

## Q1. Z-Test – Left-Tailed

A health report says college students sleep **7 hours** per night on average, with a known population standard deviation of **σ = 1.2 hours**.
Using `sleep_hours` for **all 40 students**, test at α = 0.05 whether students sleep **less than 7 hours** on average.

**Steps:** (a) write $H_0$, $H_1$ and the tail, (b) find $n$ and $\bar{x}$, (c) compute $z$, (d) find the critical value and the p-value, (e) make a decision and write a conclusion.

*(15 marks)*

In [6]:
# ===================== Q1 – YOUR CODE HERE =====================
# Z-test, left-tailed: column 'sleep_hours', mu_0 = 7, sigma = 1.2, all 40 students
data = df['sleep_hours']

mu_0 = 7
sigma = 1.2
n = len(data)
x_bar = data.mean()

z = (x_bar - mu_0) / (sigma / np.sqrt(n))

p_value = stats.norm.cdf(z)

critical_value = stats.norm.ppf(alpha)

print("Q1 - Z-Test Left-Tailed")
print(f"n = {n}")
print(f"Sample mean = {x_bar:.4f}")
print(f"Z-statistic = {z:.4f}")
print(f"Critical value = {critical_value:.4f}")
print(f"p-value = {p_value:.4f}")

if p_value <= alpha:
    print("Decision: Reject H0")
else:
    print("Decision: Fail to reject H0")


Q1 - Z-Test Left-Tailed
n = 40
Sample mean = 6.6475
Z-statistic = -1.8578
Critical value = -1.6449
p-value = 0.0316
Decision: Reject H0


**Your answer – Q1**

- **$H_0$:**
- **$H_1$:**
- **Tail (left / right / two):**
- **Test statistic:**
- **Critical value:**
- **p-value:**
- **Decision (Reject $H_0$ / Fail to reject $H_0$):**
- **Conclusion in words:**

## Q2. Z-Test – Right-Tailed

A national survey says young adults spend **5 hours** a day on screens, with a known population standard deviation of **σ = 1.5 hours**.
Using `screen_time` for **all 40 students**, test at α = 0.05 whether these students' screen time is **more than 5 hours** a day.

**Steps:** (a) write $H_0$, $H_1$ and the tail, (b) find $n$ and $\bar{x}$, (c) compute $z$, (d) find the critical value and the p-value, (e) make a decision and write a conclusion.

*(15 marks)*

In [7]:
# ===================== Q2 – YOUR CODE HERE =====================
# Z-test, right-tailed: column 'screen_time', mu_0 = 5, sigma = 1.5, all 40 students
data = df['screen_time']

mu_0 = 5
sigma = 1.5
n = len(data)
x_bar = data.mean()

z = (x_bar - mu_0) / (sigma / np.sqrt(n))

p_value = 1 - stats.norm.cdf(z)

critical_value = stats.norm.ppf(1 - alpha)

print("Q2 - Z-Test Right-Tailed")
print(f"n = {n}")
print(f"Sample mean = {x_bar:.4f}")
print(f"Z-statistic = {z:.4f}")
print(f"Critical value = {critical_value:.4f}")
print(f"p-value = {p_value:.4f}")

if p_value <= alpha:
    print("Decision: Reject H0")
else:
    print("Decision: Fail to reject H0")

Q2 - Z-Test Right-Tailed
n = 40
Sample mean = 5.4750
Z-statistic = 2.0028
Critical value = 1.6449
p-value = 0.0226
Decision: Reject H0


**Your answer – Q2**

- **$H_0$:**
- **$H_1$:**
- **Tail (left / right / two):**
- **Test statistic:**
- **Critical value:**
- **p-value:**
- **Decision (Reject $H_0$ / Fail to reject $H_0$):**
- **Conclusion in words:**

## Q3. Z-Test – Two-Tailed

Over many years, the course's final exam has had a mean of **65** with a known standard deviation of **σ = 10**.
Using `exam_score` for **all 40 students**, test at α = 0.05 whether this year's mean score is **different from 65**.

**Steps:** (a) write $H_0$, $H_1$ and the tail, (b) find $n$ and $\bar{x}$, (c) compute $z$, (d) find the critical value and the p-value, (e) make a decision and write a conclusion.

*(15 marks)*

In [8]:
# ===================== Q3 – YOUR CODE HERE =====================
# Z-test, two-tailed: column 'exam_score', mu_0 = 65, sigma = 10, all 40 students
data = df['exam_score']

mu_0 = 65
sigma = 10
n = len(data)
x_bar = data.mean()

z = (x_bar - mu_0) / (sigma / np.sqrt(n))

p_value = 2 * (1 - stats.norm.cdf(abs(z)))

critical_value = stats.norm.ppf(1 - alpha / 2)

print("Q3 - Z-Test Two-Tailed")
print(f"n = {n}")
print(f"Sample mean = {x_bar:.4f}")
print(f"Z-statistic = {z:.4f}")
print(f"Critical value = ±{critical_value:.4f}")
print(f"p-value = {p_value:.4f}")

if p_value <= alpha:
    print("Decision: Reject H0")
else:
    print("Decision: Fail to reject H0")

Q3 - Z-Test Two-Tailed
n = 40
Sample mean = 64.2000
Z-statistic = -0.5060
Critical value = ±1.9600
p-value = 0.6129
Decision: Fail to reject H0


**Your answer – Q3**

- **$H_0$:**
- **$H_1$:**
- **Tail (left / right / two):**
- **Test statistic:**
- **Critical value:**
- **p-value:**
- **Decision (Reject $H_0$ / Fail to reject $H_0$):**
- **Conclusion in words:**

---
# Part B – t-Tests (σ unknown, use one section only)

Select one section like this: `df[df['section'] == 'B']['attendance']`

## Q4. t-Test – Left-Tailed

The class teacher of **Section B** believes her students' attendance is **below 83 %**. The population standard deviation is **not known**.
Using `attendance` for **Section B only**, test at α = 0.05 whether the mean attendance is **less than 83 %**.

**Steps:** (a) write $H_0$, $H_1$ and the tail, (b) select the rows for the section, then find $n$, $\bar{x}$, $s$ and $df$, (c) compute $t$, (d) find the critical value and the p-value, (e) make a decision and write a conclusion, (f) check your answer with `stats.ttest_1samp(data, popmean, alternative=...)`.

*(15 marks)*

In [9]:
# ===================== Q4 – YOUR CODE HERE =====================
# t-test, left-tailed: column 'attendance', Section B only, mu_0 = 83, sigma unknown
data = df[df['section'] == 'B']['attendance']

mu_0 = 83

n = len(data)
x_bar = data.mean()
s = data.std()
dof = n - 1

t = (x_bar - mu_0) / (s / np.sqrt(n))

p_value = stats.t.cdf(t, dof)

critical_value = stats.t.ppf(alpha, dof)

print("Q4 - t-Test Left-Tailed")
print(f"Section = B")
print(f"n = {n}")
print(f"Sample mean = {x_bar:.4f}")
print(f"Sample standard deviation = {s:.4f}")
print(f"Degrees of freedom = {dof}")
print(f"t-statistic = {t:.4f}")
print(f"Critical value = {critical_value:.4f}")
print(f"p-value = {p_value:.4f}")

if p_value <= alpha:
    print("Decision: Reject H0")
else:
    print("Decision: Fail to reject H0")

# Check using scipy
check = stats.ttest_1samp(data, popmean=mu_0, alternative='less')
print(f"Scipy check: t = {check.statistic:.4f}, p = {check.pvalue:.4f}")

Q4 - t-Test Left-Tailed
Section = B
n = 13
Sample mean = 80.6154
Sample standard deviation = 8.3420
Degrees of freedom = 12
t-statistic = -1.0307
Critical value = -1.7823
p-value = 0.1615
Decision: Fail to reject H0
Scipy check: t = -1.0307, p = 0.1615


**Your answer – Q4**

- **$H_0$:**
- **$H_1$:**
- **Tail (left / right / two):**
- **Test statistic:**
- **Critical value:**
- **p-value:**
- **Decision (Reject $H_0$ / Fail to reject $H_0$):**
- **Conclusion in words:**

## Q5. t-Test – Right-Tailed

The college recommends at least **2.5 hours** of self-study per day. The teacher of **Section A** claims her students study **more than 2.5 hours** a day on average. σ is **not known**.
Using `study_hours` for **Section A only**, test the claim at α = 0.05.

**Steps:** (a) write $H_0$, $H_1$ and the tail, (b) select the rows for the section, then find $n$, $\bar{x}$, $s$ and $df$, (c) compute $t$, (d) find the critical value and the p-value, (e) make a decision and write a conclusion, (f) check your answer with `stats.ttest_1samp(data, popmean, alternative=...)`.

*(15 marks)*

In [10]:
# ===================== Q5 – YOUR CODE HERE =====================
# t-test, right-tailed: column 'study_hours', Section A only, mu_0 = 2.5, sigma unknown
data = df[df['section'] == 'A']['study_hours']

mu_0 = 2.5

n = len(data)
x_bar = data.mean()
s = data.std()
dof = n - 1

t = (x_bar - mu_0) / (s / np.sqrt(n))

p_value = 1 - stats.t.cdf(t, dof)

critical_value = stats.t.ppf(1 - alpha, dof)

print("Q5 - t-Test Right-Tailed")
print(f"Section = A")
print(f"n = {n}")
print(f"Sample mean = {x_bar:.4f}")
print(f"Sample standard deviation = {s:.4f}")
print(f"Degrees of freedom = {dof}")
print(f"t-statistic = {t:.4f}")
print(f"Critical value = {critical_value:.4f}")
print(f"p-value = {p_value:.4f}")

if p_value <= alpha:
    print("Decision: Reject H0")
else:
    print("Decision: Fail to reject H0")

# Check using scipy
check = stats.ttest_1samp(data, popmean=mu_0, alternative='greater')
print(f"Scipy check: t = {check.statistic:.4f}, p = {check.pvalue:.4f}")

Q5 - t-Test Right-Tailed
Section = A
n = 14
Sample mean = 2.9286
Sample standard deviation = 0.7569
Degrees of freedom = 13
t-statistic = 2.1185
Critical value = 1.7709
p-value = 0.0270
Decision: Reject H0
Scipy check: t = 2.1185, p = 0.0270


**Your answer – Q5**

- **$H_0$:**
- **$H_1$:**
- **Tail (left / right / two):**
- **Test statistic:**
- **Critical value:**
- **p-value:**
- **Decision (Reject $H_0$ / Fail to reject $H_0$):**
- **Conclusion in words:**

## Q6. t-Test – Two-Tailed

The transport office plans the college bus timetable assuming an average one-way commute of **35 minutes**. σ is **not known**.
Using `commute_min` for **Section C only**, test at α = 0.05 whether the mean commute time is **different from 35 minutes**.

**Steps:** (a) write $H_0$, $H_1$ and the tail, (b) select the rows for the section, then find $n$, $\bar{x}$, $s$ and $df$, (c) compute $t$, (d) find the critical value and the p-value, (e) make a decision and write a conclusion, (f) check your answer with `stats.ttest_1samp(data, popmean, alternative=...)`.

*(15 marks)*

In [11]:
# ===================== Q6 – YOUR CODE HERE =====================
# t-test, two-tailed: column 'commute_min', Section C only, mu_0 = 35, sigma unknown
data = df[df['section'] == 'C']['commute_min']

mu_0 = 35

n = len(data)
x_bar = data.mean()
s = data.std()
dof = n - 1

t = (x_bar - mu_0) / (s / np.sqrt(n))

p_value = 2 * (1 - stats.t.cdf(abs(t), dof))

critical_value = stats.t.ppf(1 - alpha / 2, dof)

print("Q6 - t-Test Two-Tailed")
print(f"Section = C")
print(f"n = {n}")
print(f"Sample mean = {x_bar:.4f}")
print(f"Sample standard deviation = {s:.4f}")
print(f"Degrees of freedom = {dof}")
print(f"t-statistic = {t:.4f}")
print(f"Critical value = ±{critical_value:.4f}")
print(f"p-value = {p_value:.4f}")

if p_value <= alpha:
    print("Decision: Reject H0")
else:
    print("Decision: Fail to reject H0")

# Check using scipy
check = stats.ttest_1samp(data, popmean=mu_0, alternative='two-sided')
print(f"Scipy check: t = {check.statistic:.4f}, p = {check.pvalue:.4f}")

Q6 - t-Test Two-Tailed
Section = C
n = 13
Sample mean = 29.2308
Sample standard deviation = 7.2704
Degrees of freedom = 12
t-statistic = -2.8611
Critical value = ±2.1788
p-value = 0.0143
Decision: Reject H0
Scipy check: t = -2.8611, p = 0.0143


**Your answer – Q6**

- **$H_0$:**
- **$H_1$:**
- **Tail (left / right / two):**
- **Test statistic:**
- **Critical value:**
- **p-value:**
- **Decision (Reject $H_0$ / Fail to reject $H_0$):**
- **Conclusion in words:**

---
## Q7. Summary Table *(10 marks)*

Fill in the table with your results, then answer the two short questions below it.

**Your answer – Q7**

| Q | Test | Tail | Statistic | Critical value | p-value | Decision |
|---|---|---|---|---|---|---|
| Q1 |Z-test|left|-1.8578 |-1.6449 |0.0316 |Reject H₀ |
| Q2 |Z-test |Right|2.0028 |+1.6449 |0.0226 |Reject H₀ |
| Q3 |Z-test |Two|-0.5060 |±1.9600 |0.6129|Fail to reject H₀ |
| Q4 |t-test |Left|-1.0307 |-1.7823 |0.1615 |Fail to reject H₀ |
| Q5 |t-test |Right|2.1185 |+1.7709 |0.0270 |Reject H₀ |
| Q6 |t-test |Two|-2.8611 |±2.1788 |0.0143|Reject H₀ |

**(a)** Why did Q1–Q3 use a Z-test, while Q4–Q6 used a t-test?
Q1-Q3 used a Z-test because the population standard deviation σ was given as known. Q4-Q6 used a t-test because the population standard deviation σ was unknown, so the sample standard deviation was used instead.
**(b)** In Q3 and Q4 you failed to reject $H_0$. Does this prove that $H_0$ is true? Explain in one or two sentences.
Explain in one or two sentences. No. Failing to reject H₀ does not prove that H₀ is true. It only means that there is not enough statistical evidence at the 5% significance level to reject H₀.

---
## Submission Guidelines

1. Complete every `YOUR CODE HERE` cell and every **Your answer** cell.
2. Before submitting, click **Kernel → Restart & Run All** (in Colab: **Runtime → Restart session and run all**). Make sure there are no errors and all outputs are visible.
3. Do not delete or change the starter cells.
4. Save the file as `RollNumber_Name_ZtTest.ipynb` (for example `23CS1042_Ananya_ZtTest.ipynb`).
5. Submit **only the `.ipynb` file** on the LMS before the deadline. PDF, HTML or `.py` files are not accepted.
6. This is an individual assignment. Do not copy code or answers from classmates.

**Marking:** Q1–Q6 are worth 15 marks each: hypotheses and tail (3), correct calculations (6), decision (3) and conclusion in words (3). Q7 is worth 10 marks. Total = 100.